packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -0,0 +1,71 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "../interface/FundingRateStoreInterface.sol";
+import "../../../common/implementation/Testable.sol";
+
+
+contract FundingRateStore is FundingRateStoreInterface, Testable {
+    struct Proposal {
+        FixedPoint.Unsigned rate;
+        uint256 time;
+        address proposer;
+    }
+
+    struct FundingRateRecord {
+        FixedPoint.Unsigned rate;
+        Proposal proposal;
+    }
+
+    enum ProposalState { None, Pending, Expired }
+
+    mapping(bytes32 => FundingRateRecord) public fundingRateRecords;
+
+    uint256 proposalLiveness;
+
+    constructor(address _timerAddress) public Testable(_timerAddress) {}
+
+    function getLatestFundingRateForIdentifier(bytes32 identifier)
+        external
+        override
+        view
+        returns (FixedPoint.Unsigned memory)
+    {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+
+        if (_getProposalState(fundingRateRecord.proposal) == ProposalState.Expired) {
```

Comment:
So do you think that this method should set a PENDING proposal as the current one? or do you think there should be a separate `set` method. Either way I think there should be some event emitted whenever a proposal changes state

Reply:
I don't know why anyone would call the set method. I basically think there should be as few methods as possible so the flow is as simple as possible:
- `propose`
- `withdraw` (or `getRewards`)
- `getFundingRateForIdentifier`
- `dispute`

These are all well-motivated and should require no outside intervention to keep things moving. My worry about sticking methods like `set` in there is that no user has a direct motivation to call it, so it will either never be called or it will create a more complicated workflow to push the contract into a state where you can call `propose`.

The only remaining concern I have is how to keep the pmp contract synced with the funding rates. The way the current funding rate store is designed, the contract only finds out about the new funding rate after the funding rate is already in place. You could imagine a scenario where the pmp is responsible for calling some `set` method, but I'm still not sure if that would have all the properties we want. Thoughts?

Reply:
So how would `withdraw` work? Would it pay the last `proposer`? 

> The only remaining concern I have is how to keep the pmp contract synced with the funding rates. The way the current funding rate store is designed, the contract only finds out about the new funding rate after the funding rate is already in place. You could imagine a scenario where the pmp is responsible for calling some set method, but I'm still not sure if that would have all the properties we want. Thoughts?

On this point, @chrismaree and I were thinking that the FundingRateApplier would call some update method on the FundingRateStore on its own update method

Reply:
> So how would withdraw work? Would it pay the last proposer?

It would pay the caller anything they're owed (for disputing, proposing, bond refunds, etc). The caller may have to specify what they want to be paid. Haven't dug into it enough yet.

> On this point, @chrismaree and I were thinking that the FundingRateApplier would call some update method on the FundingRateStore on its own update method

Yeah, that could make sense. What properties do you think that would have wrt the up-to-date-ness of the funding rate computations in the contract. I think we should consider a few example scenarios:
- Two proposals go through without anyone touching the pmp. Someone uses a UI to read the "current value" of each token.
- A single proposal is sent out, expires (approved), but nobody touches the funding rate store for multiple days while the pmp is used heavily.

There may be other instructive examples, but I think we should use a few examples to outline the properties/invariants that we care about and those we don't (or are willing to sacrifice).

Reply:
We could add a public view only method that reads out the effective funding rate (i.e. the current one, or an expired one that would be processed on any contract interaction)?